cmd/embedded-cluster/reset.go

```diff
@@ -0,0 +1,299 @@
+	}
+	etcd := etcdMembers{}
+	err = json.Unmarshal(out, &etcd)
+	if err != nil {
+		return false, "", fmt.Errorf("unable to unmarshal etcd member list, %w, %s", err, out)
+	}
+	if len(etcd.Members) < 3 {
+		return true, "", nil
+	}
+	if len(etcd.Members) == 3 {
+		return false, "cluster has 3 control-plane nodes, removing this node will cause etcd to lose quorum", nil
+	}
+	if len(etcd.Members)%2 != 0 {
+		return false, "cluster would have even number of control-plane nodes after resetting this node, this could cause etcd to become unstable", nil
+	}
+	return true, "", nil
+}
+
+// leaveEtcdcluster uses k0s to attempt to leave the etcd cluster
+func (h *hostInfo) leaveEtcdcluster() error {
+	out, err := exec.Command(k0s, "etcd", "leave").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("unable to leave etcd cluster: %w, %s", err, string(out))
+	}
+	return nil
+}
+
+// stopK0s attempts to stop the k0s service
+func (h *hostInfo) stopAndResetK0s() error {
+	out, err := exec.Command(k0s, "stop").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("could not stop k0s service: %w, %s", err, string(out))
+	}
+	out, err = exec.Command(k0s, "reset").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("could not reset k0s: %w, %s", err, string(out))
+	}
+	fmt.Println("Node has been reset, please reboot to ensure transient configuration is also reset")
+	return nil
+}
+
+// newHostInfo returns a populated hostInfo struct
+func newHostInfo(ctx context.Context) (hostInfo, error) {
+	currentHost := hostInfo{}
+	// populate hostname
+	err := currentHost.getHostName()
+	if err != nil {
+		return currentHost, err
+	}
+	// set up kube client
+	err = currentHost.configureKubernetesClient()
+	if err != nil {
+		return currentHost, err
+	}
+	// fetch node object
+	err = currentHost.getNodeObject(ctx)
+	if err != nil {
+		return currentHost, err
+	}
+	// control plane only stff
+	if currentHost.isControlPlane() {
+		// fetch controlNode
+		err := currentHost.getControlNodeObject(ctx)
+		if err != nil {
+			return currentHost, err
+		}
+	}
+	return currentHost, nil
+}
+
+func checkErrPrompt(err error) bool {
+	if err == nil {
+		return true
+	}
+	fmt.Println("-----")
+	fmt.Println(err)
+	fmt.Println("-----")
+	fmt.Println("An error has occured while trying to reset this node.")
+	fmt.Println("Continuing may leave the cluster in an unexpected state")
+	return prompts.New().Confirm("Do you want to continue anyway?", false)
+}
+
+var resetCommand = &cli.Command{
+	Name: "reset",
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:   "yes-really-reset",
+			Hidden: true,
+			Value:  false,
+		},
+	},
+	Usage: "Reset the node this command is run from",
+	Action: func(c *cli.Context) error {
+
+		fmt.Println("This command will completely reset this node, removing it from the cluster")
+		if !prompts.New().Confirm("Do you want to continue?", false) {
+			fmt.Println("aborting.")
+			return nil
+		}
+
+		fmt.Println("gathering facts...")
```

Comment:
I don't know that this line adds much other than cluttering the output. As I recall in the demo, we reached the next step pretty quickly.